openhcl/page_pool_alloc/src/lib.rs

```diff
@@ -23,6 +23,261 @@ use std::sync::Arc;
+    #[mesh(package = "openhcl.pagepool")]
+    struct MemoryRangeWithNode {
+        #[mesh(1)]
+        range: MemoryRange,
+        #[mesh(2)]
+        vnode: u32,
+    }
+
+    #[derive(Protobuf)]
+    #[mesh(package = "openhcl.pagepool")]
+    enum DeviceIdState {
+        #[mesh(1)]
+        Used(String),
+        #[mesh(2)]
+        Unassigned(String),
+        #[mesh(3)]
+        Leaked(String),
+    }
+
+    impl From<DeviceId> for DeviceIdState {
+        fn from(state: DeviceId) -> Self {
+            match state {
+                DeviceId::PendingRestore(name) => {
+                    panic!("should not save PendingRestore, device name: {name}")
+                }
+                DeviceId::Leaked(name) => DeviceIdState::Leaked(name),
+                DeviceId::Unassigned(name) => DeviceIdState::Unassigned(name),
+                DeviceId::Used(name) => DeviceIdState::Used(name),
+            }
+        }
+    }
+
+    impl From<DeviceIdState> for DeviceId {
+        fn from(state: DeviceIdState) -> Self {
+            match state {
+                DeviceIdState::Used(name) => DeviceId::PendingRestore(name),
+                DeviceIdState::Unassigned(name) => DeviceId::Unassigned(name),
+                DeviceIdState::Leaked(name) => DeviceId::Leaked(name),
+            }
+        }
+    }
+
+    /// The saved state for [`PagePool`].
+    #[derive(Protobuf, SavedStateRoot)]
+    #[mesh(package = "openhcl.pagepool")]
+    pub struct PagePoolState {
+        #[mesh(1)]
+        state: Vec<InnerState>,
+        #[mesh(2)]
+        device_ids: Vec<DeviceIdState>,
+        #[mesh(3)]
+        ranges: Vec<MemoryRangeWithNode>,
+    }
+
+    impl SaveRestore for PagePool {
+        type SavedState = PagePoolState;
+
+        fn save(&mut self) -> Result<Self::SavedState, vmcore::save_restore::SaveError> {
+            let state = self.inner.lock();
+            Ok(PagePoolState {
+                state: state.state.iter().map(|s| s.clone().into()).collect(),
+                device_ids: state
+                    .device_ids
+                    .iter()
+                    .map(|id| id.clone().into())
+                    .collect(),
+                ranges: self
+                    .ranges
+                    .iter()
+                    .map(|range| MemoryRangeWithNode {
+                        range: range.range,
+                        vnode: range.vnode,
+                    })
+                    .collect(),
+            })
+        }
+
+        fn restore(
+            &mut self,
+            state: Self::SavedState,
+        ) -> Result<(), vmcore::save_restore::RestoreError> {
+            // Verify that the pool describes the same regions of memory as the
+            // saved state.
+            for (current, saved) in self.ranges.iter().zip(state.ranges.iter()) {
+                if current.range != saved.range || current.vnode != saved.vnode {
+                    // TODO: return unmatched range or vecs?
```

Comment:
I think it is fatal error

Reply:
We didn't commit the restore so the caller will probably treat it as a fatal error when we return the error